// @antv/x6@2.18.1/es/view/cache.mjs
import '../../../../../../@antv_x6-common@2.0.16/node_modules/@antv/x6-common/es/index.mjs';
import { Util } from '../util/index.mjs';
import { Dictionary } from '../../../../../../@antv_x6-common@2.0.16/node_modules/@antv/x6-common/es/dictionary/dictionary.mjs';
import { getTransformToParentElement } from '../../../../../../@antv_x6-common@2.0.16/node_modules/@antv/x6-common/es/dom/transform.mjs';
import { createSVGMatrix } from '../../../../../../@antv_x6-common@2.0.16/node_modules/@antv/x6-common/es/dom/matrix.mjs';

class Cache {
    constructor(view) {
        this.view = view;
        this.clean();
    }
    clean() {
        if (this.elemCache) {
            this.elemCache.dispose();
        }
        this.elemCache = new Dictionary();
        this.pathCache = {};
    }
    get(elem) {
        const cache = this.elemCache;
        if (!cache.has(elem)) {
            this.elemCache.set(elem, {});
        }
        return this.elemCache.get(elem);
    }
    getData(elem) {
        const meta = this.get(elem);
        if (!meta.data) {
            meta.data = {};
        }
        return meta.data;
    }
    getMatrix(elem) {
        const meta = this.get(elem);
        if (meta.matrix == null) {
            const target = this.view.container;
            meta.matrix = getTransformToParentElement(elem, target);
        }
        return createSVGMatrix(meta.matrix);
    }
    getShape(elem) {
        const meta = this.get(elem);
        if (meta.shape == null) {
            meta.shape = Util.toGeometryShape(elem);
        }
        return meta.shape.clone();
    }
    getBoundingRect(elem) {
        const meta = this.get(elem);
        if (meta.boundingRect == null) {
            meta.boundingRect = Util.getBBoxV2(elem);
        }
        return meta.boundingRect.clone();
    }
}

export { Cache };
