// @antv/x6@2.18.1/es/registry/grid/index.mjs
import '../../../../../../../@antv_x6-common@2.0.16/node_modules/@antv/x6-common/es/index.mjs';
import { Registry } from '../registry.mjs';
import * as main from './main.mjs';
import { createSvgDocument, createSvgElement } from '../../../../../../../@antv_x6-common@2.0.16/node_modules/@antv/x6-common/es/dom/elem.mjs';
import { Vector } from '../../../../../../../@antv_x6-common@2.0.16/node_modules/@antv/x6-common/es/vector/index.mjs';

class Grid {
    constructor() {
        this.patterns = {};
        this.root = Vector.create(createSvgDocument(), {
            width: '100%',
            height: '100%',
        }, [createSvgElement('defs')]).node;
    }
    add(id, elem) {
        const firstChild = this.root.childNodes[0];
        if (firstChild) {
            firstChild.appendChild(elem);
        }
        this.patterns[id] = elem;
        Vector.create('rect', {
            width: '100%',
            height: '100%',
            fill: `url(#${id})`,
        }).appendTo(this.root);
    }
    get(id) {
        return this.patterns[id];
    }
    has(id) {
        return this.patterns[id] != null;
    }
}
(function (Grid) {
    Grid.presets = main;
    Grid.registry = Registry.create({
        type: 'grid',
    });
    Grid.registry.register(Grid.presets, true);
})(Grid || (Grid = {}));

export { Grid };
