// @antv/x6-common@2.0.16/es/dom/main.mjs
export { CASE_SENSITIVE_ATTR, attr, getAttribute, kebablizeAttrs, mergeAttrs, qualifyAttr, removeAttribute, setAttribute, setAttributes, styleToObject } from './attr.mjs';
export { after, append, appendTo, before, children, contains, createElement, createElementNS, createSvgDocument, createSvgElement, empty, ensureId, find, findOne, findParentByClass, index, isElement, isHTMLElement, isSVGGraphicsElement, ns, parseXML, prepend, remove, svgVersion, tagName, uniqueId } from './elem.mjs';
export { addClass, getClass, hasClass, removeClass, toggleClass } from './class.mjs';
export { getComputedStyle, hasScrollbars, setPrefixedStyle } from './style.mjs';
export { getVendorPrefixedName } from './prefix.mjs';
export { clearSelection } from './selection.mjs';
export { computeStyle, computeStyleInt, css, isCSSVariable } from './css.mjs';
export { data, getData, setData } from './data.mjs';
export { prop } from './prop.mjs';
export { breakText, measureText, splitTextByLength, text } from './text.mjs';
export { KAPPA, circleToPathData, createSlicePathData, ellipseToPathData, getPointsFromSvgElement, lineToPathData, polygonToPathData, polylineToPathData, rectToPathData, rectangleToPathData, sample, toPath, toPathData } from './path.mjs';
export { createSVGMatrix, createSVGPoint, createSVGTransform, decomposeMatrix, matrixToRotation, matrixToScale, matrixToTransformString, matrixToTranslation, parseTransformString, transformStringToMatrix } from './matrix.mjs';
export { getTransformToElement, getTransformToParentElement, rotate, scale, toLocalPoint, transform, translate } from './transform.mjs';
import './event/index.mjs';
export { MouseWheelHandle } from './mousewheel.mjs';
export { height, offset, position, width } from './position.mjs';
export { EventHook } from './event/hook.mjs';
export { Event } from './event/main.mjs';
export { EventObject } from './event/object.mjs';
