// @antv/x6-common@2.0.16/es/string/format.mjs
import { camelCase, startCase, upperCase, lowerCase, upperFirst } from 'lodash-es';
export { camelCase, lowerFirst, upperFirst } from 'lodash-es';

// @see: https://medium.com/@robertsavian/javascript-case-converters-using-lodash-4f2f964091cc
const cacheStringFunction = (fn) => {
    const cache = Object.create(null);
    return ((str) => {
        const hit = cache[str];
        return hit || (cache[str] = fn(str));
    });
};
const kebabCase = cacheStringFunction((s) => s.replace(/\B([A-Z])/g, '-$1').toLowerCase());
const pascalCase = cacheStringFunction((s) => startCase(camelCase(s)).replace(/ /g, ''));
const constantCase = cacheStringFunction((s) => upperCase(s).replace(/ /g, '_'));
const dotCase = cacheStringFunction((s) => lowerCase(s).replace(/ /g, '.'));
const pathCase = cacheStringFunction((s) => lowerCase(s).replace(/ /g, '/'));
const sentenceCase = cacheStringFunction((s) => upperFirst(lowerCase(s)));
const titleCase = cacheStringFunction((s) => startCase(camelCase(s)));

export { constantCase, dotCase, kebabCase, pascalCase, pathCase, sentenceCase, titleCase };
